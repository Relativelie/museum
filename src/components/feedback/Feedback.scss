@import "../../variables";

.feedback {
    padding: 50px $defaultMargin;
    background-color: $bgLight;
}

// TITLE AND DESCRIPTION
.feedback_title {
    @include titles($titleColor: $fontColorDark);
    margin: 0 0 30px 0;
}

.feedback_description {
    @include textSettings($fontSize: 16px, $fontColor: rgba(0, 0, 0, 0.6), $lineHeight: 20px);
    margin: 0 0 20px 0;
}

// FORM
.feedback_form {
    display: grid;
    grid-template-columns: repeat(auto-fit, min(90%, 400px));
    max-width: 850px;
    grid-gap: 20px;
}

.incorrectInputValue {
    outline-color: red !important;
    border: 1px solid red !important;
}

.feedback_input {
    @include textSettings($fontSize: 14px, $fontColor: $fontColorDark, $lineHeight: 20px);
    @include feedbackButtonAndInput;
    // width: 400px;
    padding: 14px 15px 16px 20px;

    &::placeholder {
        color: rgba(0, 0, 0, 0.6);
    }
}

.feedback_input:focus {
    border: 1px solid #000000;
    outline-color: #000000;
}

// SUBSCRIBE BUTTON
.feedback_subscribe {
    @include textSettings($fontSize: 16px, $fontColor: $fontColorDark,  $lineHeight: 20px);
    @include feedbackButtonAndInput;
    width: 141px;
    margin-top: 20px;
    background-color: transparent;
    transition: background-color .7s ease;

    &:hover {
        background-color: #75FF72;
    }
}