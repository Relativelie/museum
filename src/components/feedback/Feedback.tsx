import { MouseEvent, useState } from 'react';
import 'regenerator-runtime/runtime';
import 'core-js/stable';

import { Date } from './date/Date';
import { Email } from './email/Email';
import { Telephone } from './telephone/Telephone';
import { Time } from './time/Time';

import './Feedback.scss';
import { LoadingSpinner } from './loadingSpinner/LoadingSpinner';

export const Feedback = () => {
    const [clarifyingText, setClarifyingText] = useState('');
    const [isLoading, setIsLoading] = useState(false);

    const sendFeedbackDetails = async (e: MouseEvent<HTMLButtonElement>) => {
        const elem = e.target as HTMLFormElement;
        if (checkValues(elem)) {
            const inputValues = {
                email: elem.form[0].value,
                tel: elem.form[1].value,
                date: elem.form[2].value,
                time: elem.form[3].value,
            };
            setIsLoading(true);
            await feedbackRequest(inputValues);
            if (elem.parentElement !== null) {
                const parentElem = elem.parentElement as HTMLFormElement;
                parentElem.reset();
            }
        } else setClarifyingText('Все поля должны быть заполнены');
        setIsLoading(false);
    };

    const checkValues = (formValues: HTMLFormElement) => {
        const checkForEmptyResults = [];
        const checkForCorrectResults = [];
        for (let i = 0; i < formValues.form.length - 1; i++) {
            checkForEmptyResults.push(checkForEmptiness(formValues.form[i].value));
            checkForCorrectResults.push(checkForCorrectValue(formValues.form[i].className));
        }
        if (checkForEmptyResults.indexOf(false) === -1
            && checkForCorrectResults.indexOf(false) === -1) {
            return true;
        } else return false;
    };

    const checkForEmptiness = (value: string) => {
        if (value.length === 0) return false;
        else return true;
    };

    const checkForCorrectValue = (value: string) => {
        if (value.indexOf('incorrectInputValue') !== -1) return false;
        else return true;
    };

    const feedbackRequest = async (values: any) => {
        try {
            const resopnse = await fetch('https://sendfeedback.free.beeceptor.com/feedback', {
                method: 'POST',
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(values),
            });
            const result = await resopnse.json();
            if (result.status === 'error') {
                setClarifyingText('Something went wrong...');
            } else setClarifyingText('Заявка отправлена');
        } catch (err) {
            setClarifyingText('Something went wrong...');
        }
    };

    return (
        <div className="feedback">
            <h3 className="feedback_title">Подпишись</h3>
            <p className="feedback_description">Отправляем анонсы новых статей, выпусков и трансляций</p>
            <form className="feedback_form" action="some url">
                <Email />
                <Date />
                <Telephone />
                <Time />
                <button
                    type="button"
                    className="feedback_subscribe"
                    onClick={(e) => sendFeedbackDetails(e)}
                >
                    Подписаться
                </button>
            </form>
            <div className="processing">
                <p className="clarifyingText">{clarifyingText}</p>
                <LoadingSpinner isLoading={isLoading} />
            </div>

        </div>
    );
};
