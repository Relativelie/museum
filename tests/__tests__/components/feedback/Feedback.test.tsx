import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import * as ShallowRenderer from 'react-test-renderer/shallow';

import { Feedback } from '../../../../src/components/feedback/Feedback';

describe('Feedback - Feedback component', () => {
    test('snapshot - Feedback component', () => {
        const renderer = ShallowRenderer.createRenderer();
        renderer.render(<Feedback />);
        const tree = renderer.getRenderOutput();
        expect(tree).toMatchSnapshot();
    });
});
