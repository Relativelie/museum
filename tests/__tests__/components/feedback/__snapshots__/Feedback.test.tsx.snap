// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback - Feedback component snapshot - Feedback component 1`] = `
<div
  className="feedback"
>
  <h3
    className="feedback_title"
  >
    Подпишись
  </h3>
  <p
    className="feedback_description"
  >
    Отправляем анонсы новых статей, выпусков и трансляций
  </p>
  <form
    action="some url"
    className="feedback_form"
  >
    <Email />
    <Date />
    <Telephone />
    <Time />
    <button
      className="feedback_subscribe"
      onClick={[Function]}
      type="button"
    >
      Подписаться
    </button>
  </form>
  <div
    className="processing"
  >
    <p
      className="clarifyingText"
    >
      
    </p>
    <LoadingSpinner
      isLoading={false}
    />
  </div>
</div>
`;
